packages/pkp-walletconnect/src/lib/pkp-walletconnect.ts

```diff
@@ -0,0 +1,638 @@
+                'UNSUPPORTED_CHAINS',
+                `${chain} is not supported`
+              ),
+            });
+            rejected = true;
+            break;
+          }
+        }
+      }
+
+      // Return if session proposal was rejected
+      if (rejected) break;
+
+      // Construct the session namespace
+      namespaces[key] = {
+        accounts,
+        chains: key.includes(':') ? [key] : chains,
+        methods: requiredNamespaces[key].methods,
+        events: requiredNamespaces[key].events,
+      };
+    }
+
+    // Reject session proposal if there are no constructed namespaces for supported chains
+    for (const chain of this.supportedChains) {
+      if (!namespaces[chain]) {
+        return await this.client.rejectSession({
+          id,
+          reason: getSdkError('UNSUPPORTED_CHAINS'),
+        });
+      }
+    }
+
+    // Approve session proposal with the constructed session namespace and given relay protocol
+    return await this.approveSession({
+      id,
+      namespaces,
+      relayProtocol: relays[0].protocol,
+    });
+  }
+
+  /**
+   * Approve a session proposal from a dapp
+   * @property {number} params.id - The session ID
+   * @property {SessionTypes.Namespaces} params.namespaces - The session namespace
+   * @property {string} [params.relayProtocol] - The relay protocol
+   *
+   * @returns { Promise<SessionTypes.Struct> } - The session data
+   */
+  public approveSession: IWeb3Wallet['approveSession'] = async (params: {
+    id: number;
+    namespaces: Record<string, SessionTypes.Namespace>;
+    relayProtocol?: string;
+  }) => {
+    this.client = this._isWalletConnectInitialized(this.client);
+    return await this.client.approveSession(params);
+  };
+
+  /**
+   * Parse and reject the session proposal
+   *
+   * @param {SignClientTypes.EventArguments['session_proposal']} proposal - The session proposal
+   * @param {ErrorResponse} [reason] - The reason for rejecting the session proposal
+   */
+  public async rejectSessionProposal(
+    proposal: SignClientTypes.EventArguments['session_proposal'],
+    reason?: ErrorResponse
+  ): Promise<void> {
+    this.client = this._isWalletConnectInitialized(this.client);
+
+    const { id } = proposal;
+    return await this.rejectSession({
+      id,
+      reason: reason || getSdkError('USER_REJECTED'),
+    });
+  }
+
+  /**
+   * Reject a session proposal from a dapp
+   * @property {number} params.id - The session ID
+   * @property {ErrorResponse} params.reason - The reason for rejecting the session proposal
+   *
+   * @returns { Promise<void> }
+   */
+  public rejectSession: IWeb3Wallet['rejectSession'] = async (params: {
+    id: number;
+    reason: ErrorResponse;
+  }) => {
+    this.client = this._isWalletConnectInitialized(this.client);
+    return await this.client.rejectSession(params);
+  };
+
+  /**
+   * Approves a session request received from a dapp, processes the request using the wallet
```

Comment:
what's the diff between a session proposal and a session request?

Reply:
session proposal occurs when a dapp wants to connect to a wallet. a session request refers to any incoming requests from the dapp after the session is established - such as signing a message with one's wallet.